docs/proposals/007-variadic-generics.md

```diff
@@ -0,0 +1,663 @@
+and `GenericTypePackParameterDecl`. When the user defines type constraints on a generic type pack parameter, we will form a `GenericTypeConstraintDecl` whose `subType` is a
+`DeclRefType` referencing the `GenericTypePackParameterDecl`.
+
+### Type Pack
+A type pack represents a pack of types. The simplest form of a type pack is a `ConcreteTypePack` that is a list of concerete type packs, such as `(int, float, float3)`.
+In a generic decl such as `void f<each T>(T v)`, `T` refers to an abstract type pack represented by the generic type pack parameter `T`. The type of parameter `v` in this case
+is a `DeclRefType(GenericTypePackParameterDecl "T")`.
+The most general case of a type pack is defined by the `expand PatternExpr` type expression. In this case, the expression will be translated into a `ExpandType`, representing a
+abstract type pack that can be evaluated by substituting all `each X` expressions in the `PatternExpr` with a corresponding element in `X`, and joining all the resulting element types
+into a type pack.
+
+In summary, a type pack can be represented by one of:
+- `ConcreteTypePack`, a simple concrete list of element types.
+- `DeclRefType(GenericTypePackParameterDecl)`, a simple reference to a generic type pack parameter.
+- `ExpandType(PatternType)`, an abstract type pack resulting from expanding and evaluating `PatternType`.
+
+### `ExpandType` and `EachType`
+The type expression `expand each T` should be translated into `ExpandType(EachType(T), T)`. Here the first argument in `ExpandType` is the `PatternType`, which is what we will
+use to expand into a concrete type pack. The second argument `T` represents all the generic type pack parameters that is being captured by `PatternType`. The reason to explicitly
+keep track of captured generic type pack parameters is to make it easy to determine the size of the type pack without having to look into `PatternType`, and to ensure we never lose
+the size info even when the pattern type itself is substituted into something that is independent of any generic type pack parameters.
+
+For example, consider the substitution process on following case:
+
+```
+typealias F<T> = int; // result of F<T> is not dependent on T.
+typealias MyPack<each T : IFoo> = expand F<each T>;
+typealias Pack3 = MyPack<float, double, void>;
+```
+
+We can know from this definition that `Pack3` should evaluate to `(int, int, int)`. But let's see step-by-step how this is done in the type system.
+
+First, `Pack3` is evaluated to `MyPack<ConcreteTypePack(float, double, void)>`. To further resolve this, we will plugin the argument `ConcreteTypePack(float, double, void)` into the
+definition of `MyPack`. The definition `expand F<each T>` is represented as:
+```
+ExpandType(
+    pattern = DeclRefType(
+        GenericAppDeclRef(F,
+            args = [EachType(DeclRefType(GenericTypeParamDecl "T"))])
```

Comment:
Should this `GenericTypeParamDecl` be 
`GenericTypePackParameterDecl` ?

Reply:
You are right.